src/compiler/utilities.ts

```diff
@@ -8588,51 +8588,176 @@ export function getSetExternalModuleIndicator(options: CompilerOptions): (file:
+        return !!(compilerOptions.preserveConstEnums || computedOptions.isolatedModules.computeValue(compilerOptions));
+    }),
+    incremental: computedCompilerOption("incremental", ["composite"], compilerOptions => {
+        return !!(compilerOptions.incremental || compilerOptions.composite);
+    }),
+    declarationMap: computedCompilerOption("declarationMap", ["declaration", "composite"], (compilerOptions): boolean => {
+        return !!(compilerOptions.declarationMap && computedOptions.declaration.computeValue(compilerOptions));
+    }),
+    allowJs: computedCompilerOption("allowJs", ["checkJs"], compilerOptions => {
+        return compilerOptions.allowJs === undefined ? !!compilerOptions.checkJs : compilerOptions.allowJs;
+    }),
+    useDefineForClassFields: computedCompilerOption("useDefineForClassFields", ["target", "module"], (compilerOptions): boolean => {
+        return compilerOptions.useDefineForClassFields === undefined
+            ? computedOptions.target.computeValue(compilerOptions) >= ScriptTarget.ES2022
+            : compilerOptions.useDefineForClassFields;
+    }),
+};
 
 /** @internal */
-export function getEmitModuleKind(compilerOptions: { module?: CompilerOptions["module"]; target?: CompilerOptions["target"]; }) {
-    return typeof compilerOptions.module === "number" ?
-        compilerOptions.module :
-        getEmitScriptTarget(compilerOptions) >= ScriptTarget.ES2015 ? ModuleKind.ES2015 : ModuleKind.CommonJS;
-}
-
+export const getEmitScriptTarget = computedOptions.target.computeValue;
 /** @internal */
-export function emitModuleKindIsNonNodeESM(moduleKind: ModuleKind) {
-    return moduleKind >= ModuleKind.ES2015 && moduleKind <= ModuleKind.ESNext;
-}
-
+export const getEmitModuleKind = computedOptions.module.computeValue;
 /** @internal */
-export function getEmitModuleResolutionKind(compilerOptions: CompilerOptions) {
-    let moduleResolution = compilerOptions.moduleResolution;
-    if (moduleResolution === undefined) {
-        switch (getEmitModuleKind(compilerOptions)) {
-            case ModuleKind.CommonJS:
-                moduleResolution = ModuleResolutionKind.Node10;
-                break;
-            case ModuleKind.Node16:
-                moduleResolution = ModuleResolutionKind.Node16;
-                break;
-            case ModuleKind.NodeNext:
-                moduleResolution = ModuleResolutionKind.NodeNext;
-                break;
-            default:
-                moduleResolution = ModuleResolutionKind.Classic;
-                break;
-        }
-    }
-    return moduleResolution;
-}
+export const getEmitModuleResolutionKind = computedOptions.moduleResolution.computeValue;
+/** @internal */
+export const getEmitModuleDetectionKind = computedOptions.moduleDetection.computeValue;
+/** @internal */
+export const getIsolatedModules = computedOptions.isolatedModules.computeValue;

```

Comment:
One advantage is that these functions now automatically say what they depend on in their parameter type. The signature of this one, for example, is `(compilerOptions: Pick<CompilerOptions, "isolatedModules" | "verbatimModuleSyntax">) => boolean`